feat: add customization compute pod node (#429)
Substra/substra-documentation: docs/source/how-to/deploying-substra/howto/customize-compute-pod-node.rst

            ##
            tolerations: []
            ## @param worker.computePod.affinity.podAffinity.requiredDuringSchedulingIgnoredDuringExecution[0].labelSelector.matchExpressions[0].key Pod affinity rule defnition.
            ## @param worker.computePod.affinity.podAffinity.requiredDuringSchedulingIgnoredDuringExecution[0].labelSelector.matchExpressions[0].operator Pod affinity rule defnition.
            ## @param worker.computePod.affinity.podAffinity.requiredDuringSchedulingIgnoredDuringExecution[0].labelSelector.matchExpressions[0].values Pod affinity rule defnition.
            ## @param worker.computePod.affinity.podAffinity.requiredDuringSchedulingIgnoredDuringExecution[0].topologyKey Pod affinity rule defnition.
            ##
            affinity:
                podAffinity:
                    requiredDuringSchedulingIgnoredDuringExecution:
                    - labelSelector:
                        matchExpressions:
                        - key: statefulset.kubernetes.io/pod-name
                        operator: In
                        values:
                        - $(POD_NAME)
                    topologyKey: kubernetes.io/hostname
            ...

> We can see that set the value empty for ``nodeSelector`` and ``tolerations`` but defines an affinity rule that would force the pod to be spawned in the same pod than the worker.

Reviewer: ```suggestion
We can see that it sets empty values for ``nodeSelector`` and ``tolerations`` but defines an affinity rule that would force the pod to be spawned in the same pod than the worker.
```

.. note::
    To allow flexibility in the way we define our ``nodeSelector``, ``affinity`` and ``tolerations``, we provide the following environment variables to use as `dependent environment variables <https://kubernetes.io/docs/tasks/inject-data-application/define-interdependent-environment-variables/>`_:
    
    - ``POD_NAME``: provide the name of the worker spawning the compute function
    - ``NODE_NAME``: provide the name of the node on which the worker spawning the compute function is

On-demand GPU
=============

> In this section, we are using as an example use a dedicated node-pool to run the functions pods, provisioning nodes only when required. This example also feature using GPU sharing.

Reviewer: ```suggestion
In this section, we are using as an example a dedicated node-pool to run the functions pods, provisioning nodes only when required. This example also features using GPU sharing.
```

Activating GPU
--------------

For the following example, we will assume that you have 2 node-pools:

- ``node-pool-default``, the node-pool without GPU
- ``node-pool-gpu``, the node-pool with gpu

For our example, we assume ``node-pool-gpu`` has the following:

- labels:

>     ``node-type=substra-tasks-gpu``
>     ``node_pool=node-pool-gpu``

Reviewer: ```suggestion
    - ``node-type=substra-tasks-gpu``
    - ``node_pool=node-pool-gpu``
```

Sharing GPU between pods
------------------------

Sharing GPU on Google Kubernetes Engine + Nvidia
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This example uses time-slicing. If this is not fitting your needs, please refer to GCP documentation page on `other GPU-slicing methods <https://cloud.google.com/kubernetes-engine/docs/how-to/nvidia-mps-gpus>`_.
We will refer to the count of pods sharing the same GPU by ``<pod_count>``.

You have to activate the following settings in your node-pool:

-  through the interface
    - Set "Activate GPU"
    - Set "GPU sharing strategy" to "Time-sharing"
    - Set the Maximum number of pods sharing a GPU to ``<pod_count>``
- through Terraform
    
    .. code::

        module "<cluster_name>" {
            source  = "terraform-google-modules/kubernetes-engine/google//modules/private-cluster"
            ...
            node_pools = [
                ...
                {
                    ...
                    gpu_sharing_strategy       = "time-sharing"
                    max_shared_clients_per_gpu = <pod_count>
                    ...
                },
                ...
            ]
        ...
        }
        

> In your value file, add the following:

Reviewer: ```suggestion
In your values file, add the following:
```

.. code::

    worker:
        computePod:
            ...
            nodeSelector:
                ...
                cloud.google.com/gke-gpu-sharing-strategy: time-sharing
                cloud.google.com/gke-max-shared-clients-per-gpu: <pod_count>

> Other providers
> ---------------
> 
> For other providers, we recommend reading directly the documentation from your provider. If you're using a Nvidia GPU, you can read the reference on sharing GPU between pods (`Time-slicing <https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/gpu-sharing.html>`_ and `Multiple instance GPU (MIG) <https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/gpu-operator-mig.html>`_)

Reviewer: ```suggestion
Other cloud providers
---------------

For other cloud providers, we recommend reading directly the documentation from your provider. If you're using a Nvidia GPU, you can read the reference on sharing GPU between pods (`Time-slicing <https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/gpu-sharing.html>`_ and `Multiple instance GPU (MIG) <https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/gpu-operator-mig.html>`_)
```

> Other graphic card providers

Reviewer: ```suggestion
Other GPU providers
```
